### ISSUE TITLE
Confusion about duplicate register callback API and when to use one or the other

### ISSUE THREAD

**Issue by bogdandrutu** [open-telemetry/opentelemetry-go#3634]
Right now there are 2 ways/APIs to register a callback associated with a Observer:

* https://pkg.go.dev/go.opentelemetry.io/otel/metric@v0.35.0/instrument#WithInt64Callback
* And RegisterCallback on https://pkg.go.dev/go.opentelemetry.io/otel/metric@v0.35.0#Meter

**Questions:**
* With the `RegisterCallback` you can stop collection, which you cannot do with the first version. Is there a reason not to allow "stopping" an Observer instrument entirely?
* When it is recommended to use one or the other?

**Comment by MrAlias** [comment]
> * With the `RegisterCallback` you can stop collection, which you cannot do with the first version. Is there a reason not to allow "stopping" an Observer instrument entirely?

This comes from the [API specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-instrument-api), it says these callbacks need to be permanently registered.

> The API MUST support creation of asynchronous instruments by passing zero or more callback functions to be permanently registered to the newly created instrument.

**Comment by MrAlias** [comment]
> * When it is recommended to use one or the other?

You touched on some of the important distinctions already, but for completeness, with the `RegisterCallback` method you can

- Register a callback after an instrument is created, which can be useful if you have a separate routine that creates an instrument than the one that defines a callback
- Register a callback for more than one instrument
- Unregister whatever you register

I think the recommendation is that if you need any of these features you should use the `RegisterCallback` method and if not, just register during creation. There is a _slight_ performance improvement for the single instrument registrations, but it isn't a large enough difference to choose one over the other IMO.

**Comment by bogdandrutu** [comment]
Java has a nice way to "close" an observe instrument, still specification compliant because it is permanently associated with the instrument. See https://javadoc.io/doc/io.opentelemetry/opentelemetry-api/latest/io/opentelemetry/api/metrics/ObservableLongCounter.html

**Comment by MrAlias** [comment]
Hmm, that's an interesting idea. I'm interested to hear what others think about adding it in the future.

**Comment by bogdandrutu** [comment]
> I'm interested to hear what others think about adding it in the future.

Wouldn't be backwards incompatible? And if you make the dev checking if implementing a Closer interface is as inefficient for the code as going via the `RegisterCallback` since I need to do one more check anyway.

**Comment by MrAlias** [comment]
Yeah, I think how we add it will depend on if it is before or after the stable release.

**Comment by MrAlias** [comment]
@bogdandrutu does `close` also stop the instrument from being observed in a multi-instrument callback?

**Comment by MrAlias** [comment]
> @bogdandrutu does `close` also stop the instrument from being observed in a multi-instrument callback?

Maybe that is what this note is describing, and it is not?

![20230131_094657](https://user-images.githubusercontent.com/5543599/215841727-113bf974-eccd-4227-a5a7-d7ac8c75bd6a.png)

**Comment by Aneurysm9** [comment]
> This comes from the [API specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-instrument-api), it says these callbacks need to be permanently registered.
> 
> > The API MUST support creation of asynchronous instruments by passing zero or more callback functions to be permanently registered to the newly created instrument.

I don't think that we can enable deregistration of these callbacks without violating the spec.  If you need to deregister a callback then the instrument and callback can be associated after instrument creation.

**Comment by bogdandrutu** [comment]
@Aneurysm9 

> The API MUST support creation of asynchronous instruments by passing zero or more callback functions to be permanently registered to the newly created instrument.

To the "newly" created instrument. But if you "close" the instrument (remove it) then they will still be permanently registered to that instrument, but the instrument is no-op, hence you removed the callbaks.

**Comment by Aneurysm9** [comment]
> To the "newly" created instrument. But if you "close" the instrument (remove it) then they will still be permanently registered to that instrument, but the instrument is no-op, hence you removed the callbacks.

Where is this in the spec?

**Comment by bogdandrutu** [comment]
What do you look for the spec? Here is the [link](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-instrument-api).

**Comment by Aneurysm9** [comment]
There is no `close` method defined for any API type in the spec.
